File: crates/defi-cli/src/commands/positions.rs
use alloy::primitives::{Address, U256};
use alloy::sol_types::SolCall;
use clap::Args;

use defi_core::error::{DefiError, Result};
use defi_core::multicall::multicall_read;
use defi_core::registry::{ProtocolCategory, Registry};

use crate::output::OutputMode;

#[derive(Args)]
pub struct PositionsArgs {
    /// Wallet address to query
    #[arg(long)]
    pub address: String,

    /// Scan specific chains only (comma-separated, default: all)
    #[arg(long)]
    pub chains: Option<String>,
}

alloy::sol! {
    interface IERC20 {
        function balanceOf(address owner) external view returns (uint256);
    }

    interface IPool {
        function getUserAccountData(address user) external view returns (
            uint256 totalCollateralBase,
            uint256 totalDebtBase,
            uint256 availableBorrowsBase,
            uint256 currentLiquidationThreshold,
            uint256 ltv,
            uint256 healthFactor
        );
    }

    interface IAaveOracle {
        function getAssetPrice(address asset) external view returns (uint256);
    }
}

pub async fn run(args: PositionsArgs, registry: &Registry, output: &OutputMode) -> Result<()> {
    let user: Address = args
        .address
        .parse()
        .map_err(|e| DefiError::InvalidParam(format!("Invalid address: {e}")))?;

    let chain_filter: Option<Vec<String>> = args
        .chains
        .as_ref()
        .map(|c| c.split(',').map(|s| s.trim().to_lowercase()).collect());

    // Determine which chains to scan
    let chain_keys: Vec<String> = if let Some(ref filter) = chain_filter {
        filter.clone()
    } else {
        registry.chains.keys().cloned().collect()
    };

    let start = std::time::Instant::now();
    let mut chain_results = Vec::new();
    let mut grand_total_usd = 0.0_f64;
    let mut total_collateral_usd = 0.0_f64;
    let mut total_debt_usd = 0.0_f64;

    // Scan each chain sequentially (each chain is one multicall)
    for chain_key in &chain_keys {
        let chain = match registry.get_chain(chain_key) {
            Ok(c) => c,
            Err(_) => continue,
        };
        let rpc = chain.effective_rpc_url();

        let mut calls: Vec<(Address, Vec<u8>)> = Vec::new();

        // Track call types
        enum CallType {
            TokenBalance { symbol: String, decimals: u8 },
            LendingPosition { protocol: String, interface: String },
            NativePrice,
        }
        let mut call_types: Vec<CallType> = Vec::new();

        // 1. Token balances
        let tokens = registry.tokens.get(chain_key).cloned().unwrap_or_default();
        for token in &tokens {
            if token.address != Address::ZERO {
                call_types.push(CallType::TokenBalance {
                    symbol: token.symbol.clone(),
                    decimals: token.decimals,
                });
                calls.push((
                    token.address,
                    IERC20::balanceOfCall { owner: user }.abi_encode(),
                ));
            }
        }

        // 2. Lending positions (aave_v3 + aave_v2 pools)
        let lending_pools: Vec<(String, Address, String)> = registry
            .get_protocols_for_chain(chain_key)
            .iter()
            .filter(|p| {
                p.category == ProtocolCategory::Lending
                    && (p.interface == "aave_v3" || p.interface == "aave_v2")
            })
            .filter_map(|p| {
                p.contracts
                    .get("pool")
                    .map(|addr| (p.name.clone(), *addr, p.interface.clone()))
            })
            .collect();

        for (name, pool, iface) in &lending_pools {
            call_types.push(CallType::LendingPosition {
                protocol: name.clone(),
                interface: iface.clone(),
            });
            calls.push((*pool, IPool::getUserAccountDataCall { user }.abi_encode()));
        }

        // 3. Native price from oracle
        let oracle_addr = registry
            .get_protocols_for_chain(chain_key)
            .iter()
            .find(|p| p.interface == "aave_v3" && p.contracts.contains_key("oracle"))
            .and_then(|p| p.contracts.get("oracle").copied());

        let wrapped_native = chain.wrapped_native_address();
        if let Some(oracle) = oracle_addr {
            call_types.push(CallType::NativePrice);
            calls.push((
                oracle,
                IAaveOracle::getAssetPriceCall {
                    asset: wrapped_native,
                }
                .abi_encode(),
            ));
        }

        if calls.is_empty() {
            continue;
        }

        // Execute multicall for this chain
        let results = match multicall_read(&rpc, calls).await {
            Ok(r) => r,
            Err(_) => continue, // Skip chains with RPC errors
        };

        // Parse native price first (last call)
        let native_price = if oracle_addr.is_some() {
            let idx = results.len() - 1;
            match &results[idx] {
                Some(b) if b.len() >= 32 => U256::from_be_slice(&b[..32]).to::<u128>() as f64 / 1e8,
                _ => 0.0,
            }
        } else {
            0.0
        };

        // Parse results
        let mut token_balances = Vec::new();
        let mut lending_positions = Vec::new();
        let mut chain_value = 0.0_f64;

        for (i, ct) in call_types.iter().enumerate() {
            match ct {
                CallType::TokenBalance { symbol, decimals } => {
                    let balance = match &results[i] {
                        Some(b) if b.len() >= 32 => U256::from_be_slice(&b[..32]),
                        _ => U256::ZERO,
                    };
                    if !balance.is_zero() {
                        let bal_f64 = balance.to::<u128>() as f64 / 10f64.powi(*decimals as i32);
                        let value_usd = estimate_token_value(symbol, bal_f64, native_price);
                        if value_usd > 0.01 {
                            chain_value += value_usd;
                            token_balances.push(serde_json::json!({
                                "symbol": symbol,
                                "balance": round4(bal_f64),
                                "value_usd": round2(value_usd),
                            }));
                        }
                    }
                }
                CallType::LendingPosition {
                    protocol,
                    interface: iface,
                } => {
                    if let Some(data) = &results[i]
                        && data.len() >= 192
                    {
                        let price_decimals: u8 = if iface == "aave_v2" { 18 } else { 8 };
                        let divisor = 10f64.powi(price_decimals as i32);
                        let collateral =
                            U256::from_be_slice(&data[0..32]).to::<u128>() as f64 / divisor;
                        let debt = U256::from_be_slice(&data[32..64]).to::<u128>() as f64 / divisor;
                        let hf_raw = U256::from_be_slice(&data[160..192]);
                        let hf = if hf_raw > U256::from(u128::MAX) {
                            None
                        } else {
                            let v = hf_raw.to::<u128>() as f64 / 1e18;
                            if v > 1e10 { None } else { Some(round2(v)) }
                        };

                        if collateral > 0.01 || debt > 0.01 {
                            let net = collateral - debt;
                            chain_value += net;
                            total_collateral_usd += collateral;
                            total_debt_usd += debt;
                            lending_positions.push(serde_json::json!({
                                "protocol": protocol,
                                "collateral_usd": round2(collateral),
                                "debt_usd": round2(debt),
                                "net_usd": round2(net),
                                "health_factor": hf,
                            }));
                        }
                    }
                }
                CallType::NativePrice => {} // already parsed above
            }
        }

        if !token_balances.is_empty() || !lending_positions.is_empty() {
            grand_total_usd += chain_value;
            chain_results.push(serde_json::json!({
                "chain": chain.name,
                "native_price_usd": round2(native_price),
                "chain_total_usd": round2(chain_value),
                "token_balances": token_balances,
                "lending_positions": lending_positions,
            }));
        }
    }

    let scan_ms = start.elapsed().as_millis();

    let result = serde_json::json!({
        "address": format!("{}", user),
        "scan_duration_ms": scan_ms,
        "chains_scanned": chain_keys.len(),
        "chains_with_positions": chain_results.len(),
        "summary": {
            "total_value_usd": round2(grand_total_usd),
            "total_collateral_usd": round2(total_collateral_usd),
            "total_debt_usd": round2(total_debt_usd),
            "net_lending_usd": round2(total_collateral_usd - total_debt_usd),
        },
        "chains": chain_results,
    });

    output.print(&result)?;
    Ok(())
}

fn estimate_token_value(symbol: &str, balance: f64, native_price: f64) -> f64 {
    let s = symbol.to_uppercase();
    if s.contains("USD") || s.contains("DAI") {
        balance
    } else if s.contains("BTC") || s.contains("FBTC") {
        balance * 75000.0 // rough BTC estimate
    } else if s == "WETH" || s == "ETH" || s == "METH" || s == "CBETH" || s == "WSTETH" {
        balance * 2350.0 // rough ETH estimate
    } else {
        balance * native_price
    }
}

fn round2(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}
fn round4(x: f64) -> f64 {
    (x * 10000.0).round() / 10000.0
}

File: ts/packages/defi-cli/src/commands/positions.ts
import type { Command } from "commander";
import type { Address, Hex } from "viem";
import { encodeFunctionData, parseAbi } from "viem";
import { Registry, ProtocolCategory, multicallRead } from "@hypurrquant/defi-core";
import type { OutputMode } from "../output.js";
import { printOutput } from "../output.js";

const ERC20_ABI = parseAbi([
  "function balanceOf(address owner) external view returns (uint256)",
]);

const POOL_ABI = parseAbi([
  "function getUserAccountData(address user) external view returns (uint256 totalCollateralBase, uint256 totalDebtBase, uint256 availableBorrowsBase, uint256 currentLiquidationThreshold, uint256 ltv, uint256 healthFactor)",
]);

const ORACLE_ABI = parseAbi([
  "function getAssetPrice(address asset) external view returns (uint256)",
]);

function round2(x: number): number {
  return Math.round(x * 100) / 100;
}

function round4(x: number): number {
  return Math.round(x * 10000) / 10000;
}

function estimateTokenValue(symbol: string, balance: number, nativePrice: number): number {
  const s = symbol.toUpperCase();
  if (s.includes("USD") || s.includes("DAI")) return balance;
  if (s.includes("BTC") || s.includes("FBTC")) return balance * 75000;
  if (["WETH", "ETH", "METH", "CBETH", "WSTETH"].includes(s)) return balance * 2350;
  return balance * nativePrice;
}

function decodeU256(data: Hex | null, offset = 0): bigint {
  if (!data || data.length < 2 + (offset + 32) * 2) return 0n;
  const hex = data.slice(2 + offset * 64, 2 + offset * 64 + 64);
  return BigInt("0x" + hex);
}

interface ChainScanResult {
  chain_name: string;
  native_price: number;
  chain_value: number;
  collateral: number;
  debt: number;
  token_balances: unknown[];
  lending_positions: unknown[];
}

async function scanSingleChain(
  chainName: string,
  rpc: string,
  user: Address,
  tokens: Array<{ address: Address; symbol: string; decimals: number }>,
  lendingPools: Array<{ name: string; pool: Address; iface: string }>,
  oracleAddr: Address | undefined,
  wrappedNative: Address,
): Promise<ChainScanResult | null> {
  const calls: Array<[Address, Hex]> = [];

  type CallType =
    | { kind: "token"; symbol: string; decimals: number }
    | { kind: "lending"; protocol: string; iface: string }
    | { kind: "native_price" };

  const callTypes: CallType[] = [];

  for (const token of tokens) {
    if (token.address !== ("0x0000000000000000000000000000000000000000" as Address)) {
      callTypes.push({ kind: "token", symbol: token.symbol, decimals: token.decimals });
      calls.push([
        token.address,
        encodeFunctionData({ abi: ERC20_ABI, functionName: "balanceOf", args: [user] }),
      ]);
    }
  }

  for (const { name, pool, iface } of lendingPools) {
    callTypes.push({ kind: "lending", protocol: name, iface });
    calls.push([
      pool,
      encodeFunctionData({ abi: POOL_ABI, functionName: "getUserAccountData", args: [user] }),
    ]);
  }

  if (oracleAddr) {
    callTypes.push({ kind: "native_price" });
    calls.push([
      oracleAddr,
      encodeFunctionData({ abi: ORACLE_ABI, functionName: "getAssetPrice", args: [wrappedNative] }),
    ]);
  }

  if (calls.length === 0) return null;

  let results: (Hex | null)[];
  try {
    results = await multicallRead(rpc, calls);
  } catch {
    return null;
  }

  const nativePrice = oracleAddr
    ? Number(decodeU256(results[results.length - 1]!)) / 1e8
    : 0;

  const tokenBalances: unknown[] = [];
  const lendingPositions: unknown[] = [];
  let chainValue = 0;
  let totalColl = 0;
  let totalDebt = 0;

  for (let i = 0; i < callTypes.length; i++) {
    const ct = callTypes[i]!;
    const data = results[i] ?? null;

    if (ct.kind === "token") {
      const balance = decodeU256(data);
      if (balance > 0n) {
        const balF64 = Number(balance) / 10 ** ct.decimals;
        const valueUsd = estimateTokenValue(ct.symbol, balF64, nativePrice);
        if (valueUsd > 0.01) {
          chainValue += valueUsd;
          tokenBalances.push({
            symbol: ct.symbol,
            balance: round4(balF64),
            value_usd: round2(valueUsd),
          });
        }
      }
    } else if (ct.kind === "lending") {
      if (data && data.length >= 2 + 192 * 2) {
        const priceDecimals = ct.iface === "aave_v2" ? 18 : 8;
        const divisor = 10 ** priceDecimals;
        const collateral = Number(decodeU256(data, 0)) / divisor;
        const debt = Number(decodeU256(data, 1)) / divisor;
        const hfRaw = decodeU256(data, 5);
        let hf: number | null = null;
        if (hfRaw <= BigInt("0xffffffffffffffffffffffffffffffff")) {
          const v = Number(hfRaw) / 1e18;
          hf = v > 1e10 ? null : round2(v);
        }

        if (collateral > 0.01 || debt > 0.01) {
          const net = collateral - debt;
          chainValue += net;
          totalColl += collateral;
          totalDebt += debt;
          lendingPositions.push({
            protocol: ct.protocol,
            collateral_usd: round2(collateral),
            debt_usd: round2(debt),
            net_usd: round2(net),
            health_factor: hf,
          });
        }
      }
    }
    // native_price call — already decoded above
  }

  if (tokenBalances.length === 0 && lendingPositions.length === 0) return null;

  return {
    chain_name: chainName,
    native_price: nativePrice,
    chain_value: chainValue,
    collateral: totalColl,
    debt: totalDebt,
    token_balances: tokenBalances,
    lending_positions: lendingPositions,
  };
}

export function registerPositions(parent: Command, getOpts: () => OutputMode): void {
  parent
    .command("positions")
    .description("Cross-chain position scanner: find all your positions everywhere")
    .requiredOption("--address <address>", "Wallet address to scan")
    .option("--chains <chains>", "Comma-separated chain names (omit for all)")
    .action(async (opts: { address: string; chains?: string }) => {
      const mode = getOpts();
      const registry = Registry.loadEmbedded();

      const user = opts.address as Address;
      if (!/^0x[0-9a-fA-F]{40}$/.test(user)) {
        printOutput({ error: `Invalid address: ${opts.address}` }, mode);
        return;
      }

      const chainFilter = opts.chains
        ? opts.chains.split(",").map((s) => s.trim().toLowerCase())
        : null;

      const chainKeys = chainFilter ?? Array.from(registry.chains.keys());

      const start = Date.now();

      // Build scan params for each chain
      const scanParams: Array<{
        chainName: string;
        rpc: string;
        tokens: Array<{ address: Address; symbol: string; decimals: number }>;
        lendingPools: Array<{ name: string; pool: Address; iface: string }>;
        oracleAddr: Address | undefined;
        wrappedNative: Address;
      }> = [];

      for (const chainKey of chainKeys) {
        let chain;
        try {
          chain = registry.getChain(chainKey);
        } catch {
          continue;
        }
        const rpc = chain.effectiveRpcUrl();
        const rawTokens = registry.tokens.get(chainKey) ?? [];
        const tokens = rawTokens.map((t) => ({
          address: t.address as Address,
          symbol: t.symbol,
          decimals: t.decimals,
        }));

        const chainProtocols = registry.getProtocolsForChain(chainKey);

        const lendingPools = chainProtocols
          .filter(
            (p) =>
              p.category === ProtocolCategory.Lending &&
              (p.interface === "aave_v3" || p.interface === "aave_v2"),
          )
          .filter((p) => p.contracts?.["pool"])
          .map((p) => ({
            name: p.name,
            pool: p.contracts!["pool"] as Address,
            iface: p.interface,
          }));

        const oracleEntry = chainProtocols.find(
          (p) => p.interface === "aave_v3" && p.contracts?.["oracle"],
        );
        const oracleAddr = oracleEntry?.contracts?.["oracle"] as Address | undefined;

        const wrappedNative = (chain.wrapped_native ?? "0x5555555555555555555555555555555555555555") as Address;

        scanParams.push({ chainName: chain.name, rpc, tokens, lendingPools, oracleAddr, wrappedNative });
      }

      // Run all chains in parallel
      const chainResultsRaw = await Promise.all(
        scanParams.map((p) =>
          scanSingleChain(p.chainName, p.rpc, user, p.tokens, p.lendingPools, p.oracleAddr, p.wrappedNative),
        ),
      );

      let grandTotalUsd = 0;
      let totalCollateralUsd = 0;
      let totalDebtUsd = 0;

      const chainResults = chainResultsRaw
        .filter((r): r is ChainScanResult => r !== null)
        .map((r) => {
          grandTotalUsd += r.chain_value;
          totalCollateralUsd += r.collateral;
          totalDebtUsd += r.debt;
          return {
            chain: r.chain_name,
            native_price_usd: round2(r.native_price),
            chain_total_usd: round2(r.chain_value),
            token_balances: r.token_balances,
            lending_positions: r.lending_positions,
          };
        })
        .sort((a, b) => b.chain_total_usd - a.chain_total_usd);

      const scanMs = Date.now() - start;

      printOutput(
        {
          address: user,
          scan_duration_ms: scanMs,
          chains_scanned: chainKeys.length,
          chains_with_positions: chainResults.length,
          summary: {
            total_value_usd: round2(grandTotalUsd),
            total_collateral_usd: round2(totalCollateralUsd),
            total_debt_usd: round2(totalDebtUsd),
            net_lending_usd: round2(totalCollateralUsd - totalDebtUsd),
          },
          chains: chainResults,
        },
        mode,
      );
    });
}
